paddle/fluid/framework/ir/shrink_assign_depvar_pass.cc

```diff
@@ -0,0 +1,130 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include <stdio.h>
+#include <algorithm>
+#include <utility>
+
+#include "paddle/fluid/framework/ir/graph.h"
+#include "paddle/fluid/framework/ir/shrink_assign_depvar_pass.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+
+std::vector<ir::Node*> GetUpstreamOps(const ir::Node* op) {
+  std::vector<ir::Node*> upstreams;
+  for (ir::Node* var : op->inputs) {
+    for (ir::Node* upstream : var->inputs) {
+      PADDLE_ENFORCE(upstream->NodeType() == ir::Node::Type::kOperation);
+      upstreams.push_back(upstream);
+    }
+  }
+  return upstreams;
+}
+
+std::unique_ptr<ir::Graph> ShrinkAssignDepvarPass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  std::vector<ir::Node*> to_remove;
+  //                       op      input var
+  std::vector<std::pair<ir::Node*, ir::Node*>> to_remove_unused_input;
+
+  for (ir::Node* n : graph->Nodes()) {
+    if (n->NodeType() == ir::Node::Type::kVariable) continue;
+    for (ir::Node* in_var : n->inputs) {
+      // FIXME(typhoonzero): shouldn't got op here?
+      if (in_var->NodeType() == ir::Node::Type::kOperation) continue;
```

Comment:
If `n->inputs` has `kOperation`, the graph must have some error, so we should use `PADDLE_ENFORCE` here.

Reply:
That's why add "FIXME" here, in the `if` block, I tried to print all `n->inputs` but there's actually no node of type "Operation". Any ideas?

Reply:
Generally, I will check the type of node before I use this node, maybe I am too cautious.

Reply:
@typhoonzero do you mean you don't see kOperation here? That's the correct behavior. I think it's safe to use PADDLE_ENFORCE here. In general, the graph should be: op->var->op->var.